### PR TITLE
initial agent

### PR DESCRIPTION
**Pull request by zhouzypaul**
finished the first version of the RL agent. pls review

### CODE REVIEW

File: agent.py [hunk 1]
```diff
@@ -1,66 +1,199 @@
...
+    def __init__(self, state_size, num_stocks, past_num, resume):
         """
         this class inherits from tf.keras.models
         a general class of policy gradient RL agents using actor-critic
+        assume this agent can manage num_stocks stocks at the same time, and the agent's actions have no impact on
+        the stock environment. When performing each action, assume the agent can only sell/buy/hold one of the
+        num_stocks stocks.
 
-        :param state_size: the size of the state space, passed in from preprocessing
+        the agent's action space is a tensor of size [3 * num_stocks], where indexes [3i, 3i+1, 3i+2] refers to the
```
Comment (seanxzhan) [inline agent.py:15]: I think this is fine for now, but are we still considering this idea? See below:

There are three available actions that are based on a tree structure. The agent would first decide whether or not to act at all. Note that not acting is equivalent to holding. Then, the agent will decide whether to sell or buy. This prevents the agent from trading all the time and losing money due to transaction fees. Note that we would have a hyperparameter that represents the maximum stock value that the agent would trade at each time step. This prevents the agent from trading a huge amount of money in order to get a reward. Correspondingly, we would limit the learnable parameters for selling and buying to be some value between -1 and 1. The value represents the percentage of maximum stock value that the agent would buy or sell. (-1 means the agent would buy 100% of maximum stock value, 1 means the agent would sell 100% of maximum stock value).

Do you think it will be hard to switch between "tensor of size [3 * num_stocks]" and the action space described above? (I'm not sure how to implement a tree-structured action space)

Comment (zhouzypaul) [reply agent.py:15]: this is why i was talking about action spaces in the gc earlier. In order for the self.loss() function to work correctly, I think we need to abide by the policy gradient thm, which requires the action space to be a tensor that is actually probabilities and sums to 1. 

I think by adding a "hold" probability for each stock, it kind of achieves the same effect as a "tree structure". The network can learn to give a high probability for "hold" indexes, therefore when we sample an action, we are more likely to hold than to buy or sell. 

But of course I'm open to change. It won't be hard to change if we can come up with an action space design

Comment (zhouzypaul) [reply agent.py:15]: but yes this is definitely a great point we need to keep thinking about this. let's maybe ask antony if we can't come up with anything.

**Review by AdamWang00 — APPROVED** [review]
amazing work
